### ISSUE TITLE
[2.5.0-2.6.0] Parsing error: Cannot read property 'parent' of null

### ISSUE THREAD

**Issue by le0nik** [typescript-eslint/typescript-eslint#1114]
Repro: https://github.com/le0nik/typescript-eslint-parsing-repro.

There is a problem with `@typescript-eslint/parser` when eslint is executed in IDE/Editor(not in CLI) when you create a new file and add an import/export with curly braces anywhere in it. It triggers the following error: `Parsing error: Cannot read property 'parent' of null` .

Stack trace:
```
TypeError: Cannot read property 'parent' of null
    at Object.convertComments (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/@typescript-eslint/typescript-estree/dist/convert-comments.js:115:31)
    at Object.astConverter (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/@typescript-eslint/typescript-estree/dist/ast-converter.js:36:46)
    at Object.parseAndGenerateServices (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/@typescript-eslint/typescript-estree/dist/parser.js:295:49)
    at Object.parseForESLint (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/@typescript-eslint/parser/dist/parser.js:55:51)
    at parse (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/eslint/lib/linter/linter.js:643:22)
    at Linter._verifyWithoutProcessors (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/eslint/lib/linter/linter.js:1093:33)
    at Linter._verifyWithConfigArray (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/eslint/lib/linter/linter.js:1236:21)
    at Linter.verify (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/eslint/lib/linter/linter.js:1191:25)
    at Linter.verifyAndFix (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/eslint/lib/linter/linter.js:1381:29)
    at verifyText (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/eslint/lib/cli-engine/cli-engine.js:231:48)
```

Affected versions of `@typescript-eslint/parser`: 2.5.0-2.6.0

Some examples:

```typescript
import { namedExport } from './util'; // error!!!

const a = {}; // no error, `{}` not in `import/export`

export const foo1 = () => 'foo'; // no error
export const foo2 = () => {}; // error!!!
export const foo3 = () => ({ a: 'b' }); // error!!!
export const foo4 = () => ([ // error!!!
  { a: 'b' }, 
])
export default { c: 'd' }; // error!!!
```
----------------------------------------------------

So in summary:
- error is only present in IDE, not in CLI
- error is only present in newly created files, not in files that already existed when project was opened in IDE(or when IDE was launched)
- error is only present when you add curly braces inside import/export

**Comment by le0nik** [comment]
Error is still present with 2.5.0. Change title to reflect it.

**Comment by bradzacher** [comment]
It's weird - sometimes I can repro this, sometimes I cannot.
For example, if I paste your first failing code example into new or old file, it works as expected.

But I have run into this sometimes in other files.
It's also annoying because ESLint swallows the exception, so it's hard to debug this.

https://github.com/eslint/eslint/blob/7c8bbe0391944e1f92e083a04715bf4b3fe6be5d/lib/linter/linter.js#L667

**Comment by le0nik** [comment]
I tried to create a minimal repro https://github.com/le0nik/typescript-eslint-parsing-repro and listed steps or reproduction in the README. I can almost consistently reproduce it(sometimes Eslint doesn't give me the error, probably because of an internal error or a race condition inside of it, so I need to rename the variable and save file again for it to work).

I've updated the original post with an interesting thing I've found:

```ts
export const foo7 = () => 'foo'; // no error
export const foo8 = () => {} ; // error!!!
export foo9 = () => ([ // error!!!
  { a: 'b' }, // if you remove this object, the error is gone
]);
```

This error ony occurs when eslint sees a block `{}` after `import` or `export` in any way, like if it's a named import or if the export anything that has `{}`.

**Comment by le0nik** [comment]
Here's an error stack I've extracted:
```
"TypeError: Cannot read property 'parent' of null
    at Object.convertComments (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/@typescript-eslint/typescript-estree/dist/convert-comments.js:115:31)
    at Object.astConverter (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/@typescript-eslint/typescript-estree/dist/ast-converter.js:36:46)
    at Object.parseAndGenerateServices (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/@typescript-eslint/typescript-estree/dist/parser.js:295:49)
    at Object.parseForESLint (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/@typescript-eslint/parser/dist/parser.js:55:51)
    at parse (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/eslint/lib/linter/linter.js:643:22)
    at Linter._verifyWithoutProcessors (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/eslint/lib/linter/linter.js:1093:33)
    at Linter._verifyWithConfigArray (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/eslint/lib/linter/linter.js:1236:21)
    at Linter.verify (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/eslint/lib/linter/linter.js:1191:25)
    at Linter.verifyAndFix (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/eslint/lib/linter/linter.js:1381:29)
    at verifyText (/Volumes/Workspace/typescript-eslint-parsing-repro/node_modules/eslint/lib/cli-engine/cli-engine.js:231:48)"
```

**Comment by bradzacher** [comment]
Hmm.. If it's that particular error, it's existed for a while, it's dodgy types.
I've had a WIP PR to fix that particular dodgy type before, but it was for an unrelated error.

I'll put up a new PR to fix that

**Comment by le0nik** [comment]
Thanks for working on this!

**Comment by G-Rath** [comment]
@bradzacher I don't know if this is related, but I don't feel I have enough to make an issue out of it:

I've got a weird somewhat related interminte bug that seems to be related to 2.5.0 as well in IntelliJ, reported [here](https://youtrack.jetbrains.com/issue/WEB-42082).

Heres a recording:

http://recordit.co/8Ku8tC2Kkk

This goes away if I don't use 2.5.0, but my brains too fried by some recent long work weeks to do my typical deep dive in trying to nail down a reproduction.

I've passed this issue on to a member of the WebStorm team, who might have a better shot at reproducing, but also feel free to @ me w/ instructions on how I could get more info for you :)

I'm on Windows 10.

**Comment by valadas** [comment]
I am having the same issue, reverting to "@typescript-eslint/parser": "2.4.0" solved it temporarily...

**Comment by zoton2** [comment]
Same issue as @G-Rath here, downgrading to 2.4.0 worked too although it still seems to break in this fashion on 2.4.0 if the root workspace folder is not the root of my project (where node_modules/eslint settings are). I didn't try to downgrade further as I didn't feel it worth my time.

**Comment by miguelyoobic95** [comment]
I can confirm I have the same issue on my end

**Comment by bradzacher** [comment]
Please try v2.6.0 (released yesterday).
There was a fix within that release which (I believe) might have tangentially fixed this issue.

**Comment by le0nik** [comment]
2.6.0 didn't fix it for me. I've updated [repro](https://github.com/le0nik/typescript-eslint-parsing-repro) to use the new version.

**Comment by rfgamaral** [comment]
@bradzacher Can we assume the real fix will come with `2.7.0`?

**Comment by bradzacher** [comment]
#1120 should have fixed the comment parsing code, which I believe was the root cause of this issue.

We have an automated release that runs every monday morning and publishes whatever is in master, so that PR will be pushed out as part of the release today (should be 2.6.1, because there were only fix commits since hte last release).

**Comment by rfgamaral** [comment]
> 
> 
> #1120 should have fixed the comment parsing code, which I believe was the root cause of this issue.
> 
> We have an automated release that runs every monday morning and publishes whatever is in master, so that PR will be pushed out as part of the release today (should be 2.6.1, because there were only fix commits since hte last release).

Cool. Thanks for the clarification on the release process :)
